src/NLog/Config/ServiceRepository.cs

```diff
@@ -116,9 +144,42 @@ private object DefaultResolveInstance(Type itemType, HashSet<Type> seenTypes)
             {
                 InternalLogger.Trace("Resolve {0} done", itemType.FullName);
             }
+
+        }
+
+        private object CreateNewInstance(CompiledConstructor compiledConstructor, HashSet<Type> seenTypes)
+        {
+            var constructorParameters = compiledConstructor.Parameters;
+            if (constructorParameters == null)
+            {
+                return compiledConstructor.Ctor(null);
+            }
+
+            seenTypes = seenTypes ?? new HashSet<Type>();
+            var parameterValues = CreateCtorParameterValues(constructorParameters, seenTypes);
+            return compiledConstructor.Ctor(parameterValues);
+        }
+
+        private object CreateFromDefaultConstructor(Type itemType, ConstructorInfo defaultConstructor)
+        {
+            var compiledCtor = ReflectionHelpers.CreateLateBoundConstructor(defaultConstructor);
+            _lateBoundMap.Add(itemType, new CompiledConstructor(compiledCtor));
+            return compiledCtor(null);
         }
 
         private object CreateFromParameterizedConstructor(Type itemType, HashSet<Type> seenTypes)
+        {
+            seenTypes = seenTypes ?? new HashSet<Type>();
+
+            var ctor = GetParameterizedConstructor(itemType);
+            var constructorParameters = ctor.GetParameters();
+            var parameterValues = CreateCtorParameterValues(constructorParameters, seenTypes);
+            var compiledParameterizedCtor = ReflectionHelpers.CreateLateBoundConstructor(ctor);
+            _lateBoundMap.Add(itemType, new CompiledConstructor(compiledParameterizedCtor, constructorParameters));
+            return compiledParameterizedCtor.Invoke(parameterValues);
```

Comment:
The invoke of the constructor should be outside the lock

Reply:
OK, so needs a refactor

Reply:
how important is this? I don't know how to refactor this because it's recursive. 

(no I don't want to drop this feature ;))

Reply:
Don't think it is important, since it is only exercised for types that doesn't have a default-constructor. But please use `Activator.CreateInstance` when having registered that the type have a default-constructor (since it performs caching of constructor etc).

Reply:
> But please use `Activator.CreateInstance` when having registered that the type have a default-constructor (since it performs caching of constructor etc).

But we do caching already? I create a `CompiledConstructor` and add it to `_lateBoundMap` for default constructors. If I replace it with `Activator.CreateInstance`, I can't add it to `_lateBoundMap` and also I can't do adding a upfront cache  (do reflection on register instead of runtime) (currently not there)

Reply:
Not sure I understand. You want to remove your own caching of compiled-methods because it is too cumbersome to implement. And then I say that is okay with me, if you just fallback to `Activator.CreateInstance` when default constructor is available.

Reply:
> . You want to remove your own caching of compiled-methods because it is too cumbersome to implement.

No I don't :). I like to merge (preferable without changes) this PR and ignore the fact the locks aren't optimal. :)

Reply:
Then I don't understand why you don't implement the method so it just inserts the type in the _lateBoundMap without calling constructor. And change the code that performs the initial _lateBoundMap-lookup to call the create-constructor-method, and then performs one more _lateBoundMap-lookup after having called the create-constructor-method.

Reply:
Because maybe i'm missing something, but that won't work as the call to "createInstance and add to latebound map" is recursive. (it's done for all parameters and all parameters of those parameters etc). I don't think we need createInstance is doing lock, unlock, lock, unlock etc.

But maybe we could merge this one and you could improve the locks? That would be really helpful for me, as I am a bit stuck here.

Reply:
Hehe. Well you know my opinion. That I don't think NLog should perform dependency logic, but should just have support for using an external service-provider-interface. But if you want to commit, then commit.

Reply:
> Hehe. Well you know my opinion. 

Yes I do 👼  And I think the current code is pretty good and really useful in a many cases. I hope you are agree with me on that.